python/yugabyte_db_thirdparty/compile_commands.py

```diff
@@ -0,0 +1,285 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations
+# under the License.
+
+import json
+import logging
+import os
+import random
+import string
+import subprocess
+import copy
+import re
+import time
+
+from collections import defaultdict
+from datetime import datetime
+
+from typing import Optional, DefaultDict, Tuple, List, Dict, cast, Union, Any, Set
+
+from yugabyte_db_thirdparty import util
+from yugabyte_db_thirdparty.custom_logging import log
+
+
+COMPILE_COMMAND_FILE_SUFFIX = '.compile_command.json'
+
+TMP_DIR_ENV_VAR_NAME = 'YB_THIRDPARTY_COMPILE_COMMANDS_TMP_DIR'
+
+SOURCE_FILE_SUFFIXES = ('.h', '.hpp', '.cc', '.c', '.cxx')
+
+# The build directory has this subdirectory where we store the final compile_commands.json file.
+COMPILE_COMMANDS_SUBDIR = 'yb_compile_commands'
+
+
+def named_path_component_re_str(group_name: str) -> str:
+    return r'(?P<%s>[^/]+)' % group_name
+
+
+BAZEL_SANDBOX_PATH_RE_PREFIX_STR = '/'.join([
+    r'^.*',
+    r'[.]cache',
+    r'.*',
+    r'sandbox',
+    r'linux-sandbox',
+    r'[0-9]+',
+    r'execroot',
+    named_path_component_re_str('main_project')
+])
+
+REL_PATH_SUFFIX_RE_STR = '(?:|/(?P<rel_path>.*))'
+
+EXTERNAL_REL_PATH_RE_STR = '/'.join([
+    'external',
+    named_path_component_re_str('external_project')
+]) + REL_PATH_SUFFIX_RE_STR + '$'
+
+BAZEL_SANDBOX_PATH_RE = re.compile(BAZEL_SANDBOX_PATH_RE_PREFIX_STR + REL_PATH_SUFFIX_RE_STR + '$')
+BAZEL_SANDBOX_EXTERNAL_PATH_RE = re.compile(
+    BAZEL_SANDBOX_PATH_RE_PREFIX_STR + '/' + EXTERNAL_REL_PATH_RE_STR)
+EXTERNAL_REL_PATH_RE = re.compile('^' + EXTERNAL_REL_PATH_RE_STR)
+
+
+def get_compile_command_path_for_output_file(
+        tmp_dir: str,
+        output_path: str) -> str:
+    return tmp_dir + '/' + output_path + COMPILE_COMMAND_FILE_SUFFIX
+
+
+def get_compile_commands_dir(build_dir: str) -> str:
+    return os.path.join(build_dir, COMPILE_COMMANDS_SUBDIR)
+
+
+def get_final_compile_commands_path(build_dir: str, raw: bool) -> str:
+    return os.path.join(
+        get_compile_commands_dir(build_dir), 'compile_commands%s.json' % ('_raw' if raw else ''))
+
+
+def should_include_compile_command(compile_command: Dict[str, Union[str, List[str]]]) -> bool:
+    file_path = cast(str, compile_command['file'])
+    file_basename = os.path.basename(file_path)
+    dir_path = cast(str, compile_command['directory'])
+    dir_basename = os.path.basename(dir_path)
+    return not file_basename == 'conftest.c' and dir_basename not in ['conftest', 'conftest.dir']
```

Comment:
What is conftest and why do we have to handle it separately? (Add a comment?)

Reply:
Added a comment.